stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java

```diff
@@ -143,107 +156,276 @@ public BStruct generateNext() {
+                            }
                         }
                         break;
                     case Types.CLOB:
                         String clobValue = SQLDatasourceUtils.getString((rs.getClob(index)));
-                        bStruct.setStringField(++stringRegIndex, clobValue);
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
+                            BRefType refValue = clobValue == null ? null : new BString(clobValue);
+                            handleUnionTypeStructField(bStruct, structFields, index, ++refRegIndex, refValue,
+                                    TypeTags.STRING_TAG);
+                        } else {
+                            if (clobValue != null) {
+                                bStruct.setStringField(++stringRegIndex, clobValue);
+                            } else {
+                                handleNilToNonNilableFieldAssignment();
+                            }
+                        }
                         break;
                     case Types.NCLOB:
-                        String nClobValue = SQLDatasourceUtils.getString(rs.getNClob(index));
-                        bStruct.setStringField(++stringRegIndex, nClobValue);
+                        String nClobValue = SQLDatasourceUtils.getString((rs.getClob(index)));
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
+                            BRefType refValue = nClobValue == null ? null : new BString(nClobValue);
+                            handleUnionTypeStructField(bStruct, structFields, index, ++refRegIndex, refValue,
+                                    TypeTags.STRING_TAG);
+                        } else {
+                            if (nClobValue != null) {
+                                bStruct.setStringField(++stringRegIndex, nClobValue);
+                            } else {
+                                handleNilToNonNilableFieldAssignment();
+                            }
+                        }
                         break;
                     case Types.DATE:
                         Date date = rs.getDate(index);
-                        int fieldType = this.type.getStructFields()[index - 1].getFieldType().getTag();
-                        if (fieldType == TypeTags.STRING_TAG) {
-                            String dateValue = SQLDatasourceUtils.getString(date);
-                            bStruct.setStringField(++stringRegIndex, dateValue);
-                        } else if (fieldType == TypeTags.STRUCT_TAG) {
-                            bStruct.setRefField(++refRegIndex, createTimeStruct(date.getTime()));
-                        } else if (fieldType == TypeTags.INT_TAG) {
-                            bStruct.setIntField(++longRegIndex, date.getTime());
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
+                            if (isNillableType(structFields, index - 1, TypeTags.STRING_TAG)) {
```

Comment:
Instead of checking the union type size and having nil in all if clauses, shall we break the logic in isNillableType method into two methods where,
1. check whether the union type size is 2 and one member is nil - call this before if-else-if ladder and throw an error.
2. check whether the member types contains the required type
This will eliminate duplicate checks while going through the cases.